1.1/validator.ttl

```diff
@@ -21,22 +21,31 @@ As of GeoSPARQL 1.1, this validator is not normaltive, only informative, however
 	] ;
 	dcterms:creator "OGC GeoSPARQL Standards Working Group" ;
 	dcterms:created "2021-06-13"^^xsd:date ;
-	dcterms:modified "2021-06-13"^^xsd:date ;
+	dcterms:modified "2021-06-22"^^xsd:date ;
 	owl:versionInfo "OGC GeoSPARQL 1.1" ;
 	owl:versionIRI <http://www.opengis.net/def/geosparql/validator/1.1> ;  # same as Spec version for now
 .
 
-<feature-hasGeometry>
+<hasGeometry-hasSerialization>
     a sh:NodeShape ;
-	sh:targetClass geo:Feature ;
+	sh:targetSubjectsOf geo:hasGeometry ;
 	sh:property [
-    	sh:path geo:hasGeometry ;
-		sh:class geo:Geometry ;
-    	sh:message "Where the predicate geo:hasGeometry is used, it must indicate a geo:Geometry instance as its range value"@en ;
+    	sh:path [ 
+			sh:alternativePath (
+				geo:hasSerialization
+				geo:asGML
+				geo:asWKT
+				geo:asGeoJSON
+				geo:asKML
+				geo:asDGGS
+			) 
+		] ;
+        sh:minCount 1 ;
```

Comment:
personally, I would also add a `sh:maxCount 1`, i.e. each geo:Geometry has only one geometry serialization. Adding another geometry serialization to the same node, might result in wrong assertions (e.g. metadata concerning CRS, accuracy, etc). Making another serialization often also introduces rounding errors and other small changes that make that it's not 100% the same geometry anymore. TBD

Reply:
>  i.e. each geo:Geometry has only one geometry serialization.

We don't mandate or even suggest that in the Spec and we do have examples of multiple serializations per `Geometry`. [I've shown such examples in Annex B](https://opengeospatial.github.io/ogc-geosparql/geosparql11/spec.html#_b1_1_3_2_a_geometry_with_multiple_serializations) and I have real-world Use Cases that require multiple serilizations, for example, storage of WKT for spatial indexing (Jena) and also as GeoJSON for fast export to web map systems. We even have WKT + DGGS serializations per `Geometry` where the WTK is easy to use with most tooling but the DGGS is super fast for intersection calculations.

> Adding another geometry serialization to the same node, might result in wrong assertions

Maybe, but that's in the camp of user beware! We can't fix people using the wrong CRS etc...

We removed the proposed `inSRS` property (domain `Geometry` so that multiple serializations with differing SRS/CRSes could exist for a single `Geometry`

Reply:
wouldn't it be conceptually cleaner if each `geo:Geometry` nodes has exactly one serialization and geometry nodes of derived geometry serializations are connected? At first sight, this would not seriously harm your perfectly valid use case (use the geometry node with WKT connected for spatial indexing and then traverse the graph by three instead of two steps towards the GeoJSON geometry)? I see direct benefits for provenance (e.g. what is the "original" geometry? What is the chain of derivations?), besides the connection of metadata. If you have both a WKT and GeoJSON with different accuracy/spatialResolution (seems also like a valid use case, no?), how will you document the accuracy/spatialResolution of both with only a single `geo:Geometry` instance (note: `geo:hasSpatialResolution` has `geo:Geometry` as domain)?

It might also be more future proof wrt to later revisions of GeoSPARQL, when CRS/SRS is actually modelled and the GeoJSON is in another CRS/SRS than the WKT geometry, but also when a single geometry consists of a combination of multiple files and thus literals (e.g. Esri Shapes, glTF, OBJ with materials and/or textures, etc), to indicate a certain binary-to-text encoding for binary geometry serializations, for indicating that a geometry is part of another larger geometry (e.g. ref by internal ID), etc. We might postpone this part of the discussion to later, i.e. close the door only when it's needed and leave it open in GeoSPARQL 1.1, but I believe we'll have to do this anyway if we're serious about supporting a broad range of geometry serializations in the future.

Reply:
> If you have both a WKT and GeoJSON with different accuracy/spatialResolution (seems also like a valid use case, no?), how will you document the accuracy/spatialResolution of both with only a single `geo:Geometry` instance (note: `geo:hasSpatialResolution` has `geo:Geometry` as domain)?

That seems like a clear case of multiple `Geometry` instances being needed. Only when geometry serialisations have exactly the same CRS and coordinates there can be multiple serialisations of a single `Geometry`. CRS + coordinates are the defining attributes of a `Geometry`.

Reply:
> wouldn't it be conceptually cleaner if each geo:Geometry nodes has exactly one serialization and geometry nodes of derived geometry serializations are connected?

Initially I disagreed with the idea of limiting to one serialization per Geometry but I'm coming around to this idea, given the arguments above, especially the provenance one.

So we see Geometry as a class to contain a serialization and other non-serealization elements (e.g. geometry "role" that I've wanted but which isn't present in GeoSPARQL 1.1) but also annotations etc.

OK, if no one else can make any strong objections, I'll convert across to this idea - one serialization per Geometry - enforced with Shapes. I will need to convert examples throughout the Spec & example files to match but I'll only do that after this Shapes PR is finalized. I've raised Issue #172.

Reply:
> OK, if no one else can make any strong objections, I'll convert across to this idea - one serialization per Geometry - enforced with Shapes.

I have to object. I really can´t see any reason why we would want to do that. A good serialisation is able to encode all relevant attributes of a geometry, most importantly its coordinates and CRS. What is wrong with having multiple serialisations that faithfully represent the atttributes of the geometry? Any provenance data that is assigned to a Geometry instance will apply to all serialisations too. Is there something wrong with that? In my opinion, that is the way it is supposed to be.

Look at at this way: We can publish our ontology as RDF/XML, Turtle or JSON-LD. Three different serialisations. But those serialisations do not change the ontology itself. They are just different encodings.

Reply:
`Any provenance data that are assigned to a Geometry instance will apply to all serialisations too.` exactly this point is the issue here. This means having exactly the same coordinates, the exact same CRS/SRS, the same binary-to-text encoding, represented accuracy, the same supporting files (textures, material files, etc), etc. 

Even when discussing relatively simple geometry serializations, we notice that geometry formats enforce certain things such as "all geometry in a single fixed coordinate system" (GeoJSON). With future geometry serializations in GeoSPARQL, we get issues of relating additional data to the pure geometry such as textures and/or material information (OBJ requires a specific additional file that's different from what glTF requires, while the latter also allows to combine it with the geometry part). Besides serializations, real world tools (geometry serializers/deserializers) also have their limits. E.g. the same geometry in Rhinoceros CAD exported to OBJ and PLY might result in a slightly different mesh (slightly different vertices, faces, accuracy on coordinates, etc). Sometimes the user has some additional export options to adjust the export, but most often a perfect match is not possible. We need to take this into account and the simplest thing to do is creating different `geo:Geometry` nodes for each serialization and provide provenance links between them if necessary.

I'm also afraid that a lot of users will read past that requirement you mention (all provenance data should apply to all serialisations), resulting in questions or even worse, unreusable datasets.

`We can publish our ontology as RDF/XML, Turtle or JSON-LD. Three different serialisations. But those serialisations do not change the ontology itself. They are just different encodings.` This is only possible because they start from the same data model, i.e. RDF which is profoundly standardized. Geometry descriptions and serializations on the counterpart can vary enormously (e.g. meshes vs NURBS, 2D vs 3D, CAD vs geospatial, etc. resulting in an incredible amount of geometry formats and applications that each are good in a specific set of tasks).

In short, there are in my opinion to many "ifs" for actually making a pattern allowing multiple serializations on a single `geo:Geometry` useful in practice, so that it cannot be widely applied. The other pattern with separate `geo:Geometry` nodes for each serialization adds a tiny bit of complexity but covers way more use cases. Should we therefore allow such a rare pattern, which again increases the possibilities of how something can be modelled (and thus queried) with GeoSPARQL, potentially leading to more confusion and lower adoption?

Reply:
> `Any provenance data that are assigned to a Geometry instance will apply to all serialisations too.` exactly this point is the issue here. This means having exactly the same coordinates, the exact same CRS/SRS, the same binary-to-text encoding,

Which binary-to-text encoding is that?

> represented accuracy, the same supporting files (textures, material files, etc), etc.

I don't think textures or materials are part of the `Geometry` concept in GeoSPARQL. Geometry is a set of coordinates and the CRS used for those coordinates. Nothing more.


> Even when discussing relatively simple geometry serializations, we notice that geometry formats enforce certain things such as "all geometry in a single fixed coordinate system" (GeoJSON).  With future geometry serializations in GeoSPARQL, we get issues of relating additional data to the pure geometry such as textures and/or material information (OBJ requires a specific additional file that's different from what glTF requires, while the latter also allows to combine it with the geometry part). 

I don't think texture or material are direct attributes of geometry. I suppose additional properties are needed to represent those things, either for the `Geometry` or the `Feature` class (material probably makes most sense as a property of `Feature`).

>Besides serializations, real world tools (geometry serializers/deserializers) also have their limits. E.g. the same geometry in Rhinoceros CAD exported to OBJ and PLY might result in a slightly different mesh (slightly different vertices, faces, accuracy on coordinates, etc). Sometimes the user has some additional export options to adjust the export, but most often a perfect match is not possible. We need to take this into account and the simplest thing to do is creating different `geo:Geometry` nodes for each serialization and provide provenance links between them if necessary.

If the coordinates change because of processing, different `Geometry` nodes are needed. But that is no reason to mandate one serialisation per geometry, I think.

> I'm also afraid that a lot of users will read past that requirement you mention (all provenance data should apply to all serialisations), resulting in questions or even worse, unreusable datasets.

We should do our best to prevent users from wrong GeoSPARQL implementations. That is why I think it is important to have clarified definitions of e.g. "Feature" and "Geometry". If we see a risk of misplacement of provenance data we could mitigate that risk by providing good examples and/or SHACL shapes.


> `We can publish our ontology as RDF/XML, Turtle or JSON-LD. Three different serialisations. But those serialisations do not change the ontology itself. They are just different encodings.` This is only possible because they start from the same data model, i.e. RDF which is profoundly standardized. Geometry descriptions and serializations on the counterpart can vary enormously (e.g. meshes vs NURBS, 2D vs 3D, CAD vs geospatial, etc. resulting in an incredible amount of geometry formats and applications that each are good in a specific set of tasks).

The scope of GeoSPARQL should be knowledge representation, not representation of the wild world of serialisations. In fact, by focusing on the essentials of spatial data, GeoSPARQL could help in reducing all complexity found in formats and applications. 

> In short, there are in my opinion to many "ifs" for actually making a pattern allowing multiple serializations on a single `geo:Geometry` useful in practice, so that it cannot be widely applied. The other pattern with separate `geo:Geometry` nodes for each serialization adds a tiny bit of complexity but covers way more use cases. Should we therefore allow such a rare pattern, which again increases the possibilities of how something can be modelled (and thus queried) with GeoSPARQL, potentially leading to more confusion and lower adoption?

I think the proposed shape is confusing. If a geometry is defined by its coordinates and CRS, why shouldn't you be able to use different serialisations if they have exactly the same coordinates and CRS?

Reply:
`Which binary-to-text encoding is that?`
Many options are available (base32, base64, base122, hex, etc) to encode a binary geometry serialization such as binary PLY, GLB, DWG, etc. I don't think we should mandate a single encoding as each of them have their advantages and disadvantages. Leaving the option open to users means that this information should somehow be connectable to the `geo:Geometry` instance, unless you make specific subproperties of a future `geo:asPly-binary` such as `geo:asPly-binary-base32`, `geo:asPly-binary-base64`, etc.

`I don't think textures or materials are part of the Geometry concept`
I certainly agree that (geometry) textures and materials are different from pure geometry, but they are closely related.

`I don't think texture or material are direct attributes of geometry.`
I believe they are. Note that there's a difference between geometry texture and geometry material of a `geo:Geometry` vs physical texture and physical material of a `geo:Feature` described. You might e.g. indicate that the geometry description of a physical chair is yellow to highlight it in a certain visualization, while the actual chair (and thus its physical material) is red.

`The scope of GeoSPARQL should be knowledge representation, not representation of the wild world of serialisations.`
But haven't we been doing this all along GeoSPARQL 1.0 and 1.1 (with good reasons)? If we want to include support for more serializations in GeoSPARQL, should we do a breakdown of the widely used geometry serializations? Don't get me wrong. I do believe there might be a place for a full RDF-based way of describing geometry, e.g. to describe a translation between two geometry serializations or for describing parametric geometry. But I don't think it will ever be as practical and widespread as common, non-RDF geometry serializations (e.g. GeoJSON, OBJ, DWG, glTF, etc.) to be used on a dataset level. Therefore, a practical approach to geometry serializations should also be possible.

Reply:
> > Which binary-to-text encoding is that?
> 
> Many options are available (base32, base64, base122, hex, etc) to encode a binary geometry serialization such as binary PLY, GLB, DWG, etc. I don't think we should mandate a single encoding as each of them have their advantages and disadvantages. Leaving the option open to users means that this information should somehow be connectable to the `geo:Geometry` instance, unless you make specific subproperties of a future `geo:asPly-binary` such as `geo:asPly-binary-base32`, `geo:asPly-binary-base64`, etc.

At the moment we don't have binary serialisations. But if we would have, and we would want to support multiple binary-to-text encodings, those should logically be properties of the serialisation, not of the Geometry.

> > I don't think textures or materials are part of the Geometry concept
> 
> I certainly agree that (geometry) textures and materials are different from pure geometry, but they are closely related.
> 
> > I don't think texture or material are direct attributes of geometry.
> 
> I believe they are. Note that there's a difference between geometry texture and geometry material of a `geo:Geometry` vs physical texture and physical material of a `geo:Feature` described. You might e.g. indicate that the geometry description of a physical chair is yellow to highlight it in a certain visualization, while the actual chair (and thus its physical material) is red.

True. But in GeoSPARQL a geometry is a set of coordinates in a CRS. Not its depiction. How to style or visualise a Geometry is not covered yet, but if it ever was (breaking separation of model and presentation), probably some intermediate construct would be needed.

> 
> > The scope of GeoSPARQL should be knowledge representation, not representation of the wild world of serialisations.
> 
> But haven't we been doing this all along GeoSPARQL 1.0 and 1.1 (with good reasons)?

Well, serialisations are included of course. They are necessary because pure RDF at the moment can be considered inadequate for representing geometries. But GeoSPARQL does not get involved with particularities of serialisations. It refers to the serialisation specifications for that.

If we want to include support for more serializations in GeoSPARQL, should we do a breakdown of the widely used geometry serializations? Don't get me wrong. I do believe there might be a place for a full RDF-based way of describing geometry, e.g. to describe a translation between two geometry serializations or for describing parametric geometry. But I don't think it will ever be as practical and widespread as common, non-RDF geometry serializations (e.g. GeoJSON, OBJ, DWG, glTF, etc.) to be used on a dataset level. Therefore, a practical approach to geometry serializations should also be possible.

Sure. Although I do hope that RDF will evolve to better support for the type of data that Geometry is. But the point remains that Geometry is a clearly defined concept in GeoSPARQL. I think there should be a clear separation between an abstract model for spatial things and ways in which they may be encoded.